code/modules/reagents/chemistry/equilibrium.dm

```diff
@@ -0,0 +1,293 @@
+		for(var/P in reaction.required_catalysts)
+			var/datum/reagent/catalyst = P
+			if(catalyst == R.type)
+				total_matching_catalysts++
+
+	if(!(total_matching_catalysts == reaction.required_catalysts.len))
+		return FALSE
+
+	//All good!
+	return TRUE
+
+/*
+* Calculates how much we're aiming to create
+*
+* Specifically calcuates multiplier, product_ratio, target_vol
+* Also checks to see if these numbers are sane, returns a TRUE/FALSE
+* Generally an internal proc
+*/
+/datum/equilibrium/proc/calculate_yield()
+	if(to_delete)
+		return FALSE
+	if(!reaction)
+		stack_trace("Tried to calculate an equlibrium for reaction [reaction.type], but there was no reaction set for the datum")
+		return FALSE
+
+	multiplier = INFINITY
+	for(var/B in reaction.required_reagents)
+		multiplier = min(multiplier, round((holder.get_reagent_amount(B) / reaction.required_reagents[B]), CHEMICAL_QUANTISATION_LEVEL))
+	
+	if(!length(reaction.results)) //Incase of no reagent product
+		product_ratio = 1
+		target_vol = INFINITY
+		for(var/B in reaction.required_reagents)
+			target_vol = min(target_vol, multiplier * reaction.required_reagents[B])
+		if(target_vol == 0 || multiplier == 0)
+			return FALSE
+		//Sanity Check
+		if(target_vol == INFINITY || multiplier == INFINITY) //I don't see how this can happen, but I'm not bold enough to let infinities roll around for free
+			to_delete = TRUE
+			CRASH("Tried to calculate target vol for [reaction.type] with no products, but could not find required reagents for the reaction. If it got here, something is really broken with the recipe.")
+		return TRUE
+
+	product_ratio = 0
+	target_vol = 0
+	for(var/P in reaction.results)
+		target_vol += (reaction.results[P]*multiplier)
+		product_ratio += reaction.results[P]
+	if(target_vol == 0 || multiplier == INFINITY)
+		return FALSE
+	return TRUE
+/*
+* Main reaction processor - Increments the reaction by a timestep
+*
+* First checks the holder to make sure it can continue
+* Then calculates the purity and volume produced.TRUE
+* Then adds/removes reagents
+* Then alters the holder pH and temperature, and calls reaction_step
+* Arguments:
+* * delta_time - the time displacement between the last call and the current
+*/
+/datum/equilibrium/proc/react_timestep(delta_time)
+	if(!calculate_yield())
+		to_delete = TRUE
+		return
+	if(!check_reagent_properties())
+		to_delete = TRUE
+		return
+	
+	var/deltaT = 0 //how far off optimal temp we care
+	var/deltapH = 0 //How far off the pH we are
+	var/cached_pH = holder.pH
+	var/cached_temp = holder.chem_temp
+	var/purity = 1 //purity of the current step
+
+	//Begin checks
+	//Calculate DeltapH (Deviation of pH from optimal)
+	//Within mid range
+	if (cached_pH >= reaction.optimal_pH_min  && cached_pH <= reaction.optimal_pH_max)
+		deltapH = 1
+	//Lower range
+	else if (cached_pH < reaction.optimal_pH_min)
+		if (cached_pH < (reaction.optimal_pH_min - reaction.determin_pH_range))
+			deltapH = 0
+			//If outside pH range, 0
```

Comment:
Put comments like these before the code they're commenting, not after.

Though in general I dislike comments that explain what, and not why/how.

Reply:
Tidied up the code around there